scalyr_agent/agent_status.py

```diff
@@ -39,7 +39,41 @@
 import six
 
 
-class AgentStatus(object):
+class BaseAgentStatus(object):
+    """
+    Base agent status class which implements "to_dict()" method.
+    """
+
+    def to_dict(self):
+        # type: () -> dict
+        """
+        Return dictionary version of the status object. This dictionary contains only simple /
+        native values and is JSON serializable.
+        """
+        result = {}
+        result = self.__dict__
+
+        # Recursively convert nested objects to dicts
+        for key, value in result.items():
+            if isinstance(value, (list, tuple)):
+                items = []
+                for item in value:
+                    if hasattr(item, "to_dict"):
+                        item = item.to_dict()
+                    items.append(item)
+                result[key] = items
+            elif isinstance(value, dict):
+                for item_key, item_value in value.items():
+                    if hasattr(item_value, "to_dict"):
+                        item_value = item_value.to_dict()
+                    result[key][item_key] = item_value
+            elif hasattr(value, "to_dict"):
+                result[key] = value.to_dict()
```

Comment:
Let's raise an exception or something if none of the `if` statements match here.. otherwise we'd be silently dropping values.

Reply:
That behavior is intentional - if no if statement matches, that value is already simple (int, string, similar) and we want to use it as-is without any formatting.

And as we manipulate copy of self.__dict__ in place and not create a new dict, that's fine.

I will also add some test cases for this functionality.